dask_kubernetes/operator/controller/controller.py

```diff
@@ -392,67 +403,92 @@ async def get_desired_workers(scheduler_service_name, namespace, logger):
-                await corev1api.create_namespaced_pod(
+@kopf.on.field("daskworkergroup.kubernetes.dask.org", field="spec.worker.replicas")
+async def daskworkergroup_replica_update(
+    name, namespace, meta, spec, new, body, logger, **kwargs
+):
+    cluster_name = spec["cluster"]
+
+    # Replica updates can come in quick succession and the changes must be applied atomically to ensure
+    # the number of workers ends in the correct state
+    async with worker_group_scale_locks[f"{namespace}/{name}"]:
+        async with kubernetes.client.api_client.ApiClient() as api_client:
+            customobjectsapi = kubernetes.client.CustomObjectsApi(api_client)
+            corev1api = kubernetes.client.CoreV1Api(api_client)
+
+            try:
+                cluster = await customobjectsapi.get_namespaced_custom_object(
+                    group="kubernetes.dask.org",
+                    version="v1",
+                    plural="daskclusters",
                     namespace=namespace,
-                    body=data,
+                    name=cluster_name,
                 )
-            logger.info(
-                f"Scaled worker group {name} up to {spec['worker']['replicas']} workers."
-            )
-        if workers_needed < 0:
-            worker_ids = await retire_workers(
-                n_workers=-workers_needed,
-                scheduler_service_name=f"{spec['cluster']}-scheduler",
-                worker_group_name=name,
+            except ApiException as e:
+                if e.status == 404:
+                    # No need to scale if worker group is deleted, pods will be cleaned up
+                    return
+                else:
+                    raise e
+
+            cluster_labels = cluster.get("metadata", {}).get("labels", {})
+
+            workers = await corev1api.list_namespaced_pod(
                 namespace=namespace,
-                logger=logger,
+                label_selector=f"dask.org/workergroup-name={name}",
             )
-            logger.info(f"Workers to close: {worker_ids}")
-            for wid in worker_ids:
-                await corev1api.delete_namespaced_pod(
-                    name=wid,
+            current_workers = len(
+                [w for w in workers.items if w.status.phase != "Terminating"]
+            )
+            desired_workers = new
+            workers_needed = desired_workers - current_workers
+            annotations = _get_dask_cluster_annotations(meta)
+            if workers_needed > 0:
+                for _ in range(workers_needed):
+                    data = build_worker_pod_spec(
+                        worker_group_name=name,
+                        namespace=namespace,
+                        cluster_name=cluster_name,
+                        uuid=uuid4().hex[:10],
+                        spec=spec["worker"]["spec"],
+                        annotations=annotations,
+                    )
+                    kopf.adopt(data, owner=body)
```

Comment:
FYI this is the line that causes pods tobe cleaned up. The `Pod` gets adopted by the `DaskWorkerGroup`, which in turn has been adopted by the `DaskCluster`. When a k8s resource is deleted it will cascade delete the child resources.

If this isn't happening that's an interesting but separate problem.

Reply:
Yeah it wasn't happening until I added the `owner=body` argument, which explicitly set the owner of the pod to the worker group.  I noticed the issue after I introduced this change and I haven't noticed issues in the current release with this, so its either that something else is causing the pods to be cleaned up (I traced through the flow and I couldn't find anything that would be cleaning pods up) or more likely `adopt` behaves differently when used in the field update callback?

Reply:
> or more likely adopt behaves differently when used in the field update callback

Yeah you're probably right, although I'm a little surprised. Explicitly setting the owner seems like a great solution though.